File: 2022/java/Day3.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Day3d {
    public static void main(String ... args) throws Exception {
        var input = Files.readString(Path.of("../input/day3.txt"));
        var sum = input.lines().mapToInt(l -> {
            var p1 = l.substring(0, l.length()/2);
            var p2 = l.substring(l.length()/2, l.length());
            return commonPriority(List.of(p1, p2));
        }).sum();
        System.out.printf("part 1: %d\n", sum);

        var counter = IntStream.range(0, input.length()).iterator();
        var sum2 = input.lines()
            .collect(Collectors.groupingBy(c -> counter.nextInt()/3)).values().stream()
            .mapToInt(l -> commonPriority(l)).sum();
        System.out.printf("part 2: %d\n", sum2); 
    }

    static int commonPriority(List<String> strings) {
        // common character in list of strings
        var c = strings.stream().reduce((s1, s2) -> commonChars(s1, s2)).get().charAt(0);
        return c < 'a' ? c - 'A' + 27 : c - 'a' + 1; // convert to priority
    }

    static String commonChars(String a, String b) {
        return a.chars().filter(i1 -> b.chars().anyMatch(i2 -> i1 == i2)).distinct()
            .mapToObj(c -> Character.toString(c)).collect(Collectors.joining());
    }
}

File: 2022/go/day3/day3.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day3.txt")
	lines := strings.Split(string(input), "\n")

	sum1 := 0
	for _, line := range lines {
		p1, p2 := []byte(line[:len(line)/2]), []byte(line[len(line)/2:])
		sum1 += priority(commonItem(p1, p2)[0])
	}
	fmt.Printf("part 1: %v\n", sum1)

	sum2 := 0
	var group [][]byte
	for _, line := range lines {
		group = append(group, []byte(line))
		if len(group) == 3 {
			sum2 += priority(commonItem(group...)[0])
			group = nil
		}
	}
	fmt.Printf("part 2: %v\n", sum2)
}

func commonItem(values ...[]byte) []byte {
	res := values[0]
	for _, next := range values[1:] {
		res = commonValues(res, next)
	}
	return res
}

func commonValues(a, b []byte) []byte {
	res := []byte{}
	for _, v := range a {
		if contains(b, v) && !contains(res, v) {
			res = append(res, v)
		}
	}
	return res
}

func contains(items []byte, v byte) bool {
	for _, v1 := range items {
		if v1 == v {
			return true
		}
	}
	return false
}

func priority(c byte) int {
	if c < 'a' {
		return int(c - 'A' + 27)
	} else {
		return int(c - 'a' + 1)
	}
}
